### ISSUE TITLE
Honor AWS EventBridge Put Limits

### ISSUE THREAD

**Issue by embano1** [vmware-samples/vcenter-event-broker-appliance#69]
Requires batching (max 10 entries per `PutRequest`) and rate limiting (per region/account) as described in the limits: https://docs.aws.amazon.com/eventbridge/latest/userguide/cloudwatch-limits-eventbridge.html

**Comment by martindekov** [comment]
Hey @embano1 for context I have couple of questions as per the current implementation:
https://github.com/vmware-samples/vcenter-event-broker-appliance/blob/4d3027881bf9c0d5de2847385006b19373fe003e/vmware-event-router/internal/processor/aws_event_bridge.go#L198

So first questions is, does events arrive in batch from vCenter. I see this line of code counts them:
https://github.com/vmware-samples/vcenter-event-broker-appliance/blob/4d3027881bf9c0d5de2847385006b19373fe003e/vmware-event-router/internal/stream/vcenter.go#L129

I though we can receive only one event form vCenter per request. I think this proves me wrong? I suppose couple of things in the vCenter can happen in parallel? So batch of events are send to the router. If this is the case we don't explicitly set max 10 entries here:
https://github.com/vmware-samples/vcenter-event-broker-appliance/blob/4d3027881bf9c0d5de2847385006b19373fe003e/vmware-event-router/internal/processor/aws_event_bridge.go#L208

So essentially we set the full stack to event bridge which we receive from vCenter is that so?

Do you suggest to spawn a goroutine which will do this:
https://github.com/vmware-samples/vcenter-event-broker-appliance/blob/4d3027881bf9c0d5de2847385006b19373fe003e/vmware-event-router/internal/processor/aws_event_bridge.go#L208

and will make sure we will send the events to event bridge at max 10 events. So:

1. 15 events from vCenter arrive
2. first 10 events are sent to event bridge
3. last 5 events are kept
4. 3 events arrive from vCenter
5. the previous 5 are appended with the current 3 so 8 events are sent to event bridge

As currently as far as I understood it we send everything we received from vCenter. Sorry for the long post I want to be explicit with the approach especially in the beginning of my contribution :smile:

**Comment by embano1** [comment]
Thank you Martin for looking into this and your contribution! All good questions, let me answer with my thoughts on how I thought about addressing this `TODO` item.

I think this issue is actually two issues (at least related ones):

1. Honor batch limits per `PutEvents` call
2. Honor region/account quotas (i.e. rate limits) per `PutEvents` call

The comments in the processor file provide some links to these topics. Here's my thoughts on the two subjects.

## Batch Limits
According to the AWS EventBridge API [docs](https://docs.aws.amazon.com/eventbridge/latest/APIReference/API_PutEvents.html) min/max for `PutEvents` API call are 1/10 array members (`Entries: []*eventbridge.PutEventsRequestEntry{}`).

Yes, events from vCenter are delivered through govmomi via `[]types.BaseEvent`. I've seen occurrences with 10 entries (not more though, but this could be a false assumption). And yes, that is the issue we need to address so we stop at max 10 entries for the `PutEvents` call. I thought about using this library (or code) for performing batch operations: https://github.com/pacedotdev/batch

My suggestions for the implementation would be:

- Don't use Go routines to not violate the ordered delivery to EventBridge we have as of today
- In case of `len(baseEvent) > 10` split them in batches in the **same** processing invocation window
- Incoming vCenter events won't be processed until we return from this function so no need to align the batches with further incoming events from vCenter
- In case of error (which is not handled by the default AWS EventBridge client configuration (see below) log it and continue with the remaining batch items or next incoming events (if no further items in batch remain)

## Region/Account limits (quotas)
According to the AWS EventBridge Quotas [doc](https://docs.aws.amazon.com/eventbridge/latest/userguide/cloudwatch-limits-eventbridge.html#putevents-limits) (as of Mar 22nd) these range from 400-2400 requests per second (RPS) depending on the region. In the current implementation (using AWS SDK Go default client) retries are transparently implemented using the [`DefaultRetryer`](https://github.com/aws/aws-sdk-go/blob/172af2f58676e3d2cc733bc01f229fa4a27d6ba4/aws/config.go#L108). Details on its configuration can be found [here](https://github.com/aws/aws-sdk-go/blob/84fbd57ef75762a07aade079776907d01be3891d/aws/client/default_retryer.go#L16). I haven't tested these defaults in practice (some max values are set to 5min) and so I don't know what happens if the client (i.e. a `PutEvents` request) gets throttled (sleep) for such a long time. Perhaps it times out and falls under our current "at most once delivery" guarantees (in case of error we log it).

We could test this using the EventBridge interface [implementation](http://github.com/aws/aws-sdk-go/blob/master/service/eventbridge/eventbridgeiface/interface.go#L17), also described in this [tutorial](https://aws.amazon.com/blogs/developer/mocking-out-then-aws-sdk-for-go-for-unit-testing/) based on SQS. This would require a minor code change on the processor to use the interface type. I will file a separate PR to move to using the interface.

**Comment by martindekov** [comment]
Thanks for the input @embano1 I will start working on it 👍
